Adds couchbase vector storage

chedim opened the pull request:
<!-- Thank you so much for your contribution! -->
<!-- Please fill in all the sections below. -->

<!-- Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples. -->
<!-- Please note that PRs with breaking changes will be rejected. -->
<!-- Please note that PRs without tests will be rejected. -->

<!-- Please note that PRs will be reviewed based on the priority of the issues they address. -->
<!-- We ask for your patience. We are doing our best to review your PR as quickly as possible. -->
<!-- Please refrain from pinging and asking when it will be reviewed. Thank you for understanding! -->


## Issue
<!-- Please specify the ID of the issue this PR is addressing. For example: "Closes #1234" -->
Closes #1481

## Change
<!-- Please describe the changes you made. -->
Adds couchbase vector storage.


## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [x] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [x] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new model integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [x] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [x] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [x] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)


## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

chedim committed:
moving to thinkphone

chedim committed:
Merge branch 'langchain4j:main' into main

chedim committed:
Adds couchbase vector storage

chedim committed:
Merge branch 'langchain4j:main' into main

chedim commented on langchain4j-couchbase/pom.xml:
```diff
@@ -0,0 +1,107 @@
+
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
```
yup, thank you

chedim commented on langchain4j-couchbase/pom.xml:
```diff
@@ -0,0 +1,107 @@
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-simple</artifactId>
```
They shouldn't be there at all, I forgot to remove those. Thank you!

chedim committed:
Cleaning up dependencies and removing mentions of filtering

chedim commented on langchain4j-couchbase/src/test/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStoreIT.java:
```diff
@@ -0,0 +1,90 @@
+import java.time.Duration;
+
+@Testcontainers
+class CouchbaseEmbeddingStoreIT extends EmbeddingStoreIT {
```
I ran into issues with filtering so, for now we decided to not implement it and release only storing and removing embeddings.

chedim commented on langchain4j-couchbase/src/test/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStoreRemoveIT.java:
```diff
@@ -0,0 +1,221 @@
+
+@Disabled
+@Testcontainers
+class CouchbaseEmbeddingStoreRemoveIT {
```
Removed it

chedim committed:
Update langchain4j-couchbase/src/main/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStore.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

chedim committed:
removes filtering

chedim committed:
addresses some PR comments

chedim commented on langchain4j-couchbase/src/main/java/dev/langchain4j/store/embedding/couchbase/CouchbaseSearchMetadataFilterMapper.java:
```diff
@@ -0,0 +1,134 @@
+
+import java.util.Collection;
+
+class CouchbaseSearchMetadataFilterMapper {
```
Because it would use sql++ filtering for removing embeddings and FTS query filters for finding relevant vectors.

chedim committed:
Merge branch 'main' into main

chedim commented on langchain4j-couchbase/src/main/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStore.java:
```diff
@@ -0,0 +1,281 @@
+
+    @Override
+    public void removeAll(Collection<String> ids) {
+        ids.forEach(collection::remove);
```
No, there's no bulk-delete method in CB Java SDK, and that's intentional as routing of the deletion request to the proper vbucket (and, therefore, cb node) is done on the client due to CB's architecture.

chedim commented:
Failing test in `main`:
```
[ERROR] dev.langchain4j.store.embedding.inmemory.InMemoryEmbeddingStoreTest.test_backwards_compatibility_with_0_27_1 -- Time elapsed: 0.012 s <<< FAILURE!
```

But I don't think its related to this PR? Please help? @langchain4j, thank you!

dliubarskyi committed:
Merge branch 'main' into main

chedim committed:
Update langchain4j-couchbase/pom.xml

Co-authored-by: LangChain4j <langchain4j@gmail.com>

Heezer commented on pom.xml:
```diff
@@ -26,6 +26,7 @@
         <module>langchain4j-bedrock</module>
         <module>langchain4j-chatglm</module>
         <module>langchain4j-cohere</module>
+        <module>langchain4j-couchbase</module>
```
wrong location, this shall be moved to the bottom list

Heezer commented on langchain4j-couchbase/src/test/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStoreIT.java:
```diff
@@ -0,0 +1,86 @@
+
+@Testcontainers
+class CouchbaseEmbeddingStoreIT extends EmbeddingStoreIT {
+    public static final Integer TEST_DIMENSIONS = 384;
```
```suggestion
   static final Integer TEST_DIMENSIONS = 384;
```

Heezer commented on langchain4j-couchbase/src/test/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStoreCloudIT.java:
```diff
@@ -0,0 +1,52 @@
+import static dev.langchain4j.internal.Utils.randomUUID;
+
+@Disabled
+@EnabledIfEnvironmentVariable(named = "CB_CLUSTER", matches = ".+")
```
`CB_CLUSTER` is a bit ambiguous, since LC4J contains a lot of other service implementations, so my proposal would be: `COUCHBASE_CLUSTER_URL`. The rest of env variables shall use the full service name as well.

Heezer commented on langchain4j-couchbase/src/test/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStoreCloudIT.java:
```diff
@@ -0,0 +1,52 @@
+
+import static dev.langchain4j.internal.Utils.randomUUID;
+
+@Disabled
```
I assume `@Disabled` is not needed, `@EnabledIfEnvironmentVariable` does enabling/disabling.

Heezer commented on langchain4j-couchbase/src/main/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStore.java:
```diff
@@ -0,0 +1,332 @@
+     * @param bucketName     name of a cluster bucket in which to store the embeddings
+     * @param scopeName      name of a scope in the bucket in which to store the embeddings
+     * @param collectionName name of a collection in the scope in which to store the embeddings
+     * @param searchIndexName    name of the FTS index to be used for searching embeddings
```
param `dimensions` is missing here

chedim committed:
Changed dependencies

chedim committed:
adds couchbase docs; bumps couchbase version

chedim committed:
Update langchain4j-couchbase/src/test/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStoreIT.java

Co-authored-by: Alexey Titov <33568148+Heezer@users.noreply.github.com>

chedim committed:
Addresses PR requests

chedim committed:
Merge branch 'main' into main

Heezer commented on pom.xml:
```diff
@@ -61,6 +61,7 @@
         <module>langchain4j-vearch</module>
         <module>langchain4j-vespa</module>
         <module>langchain4j-weaviate</module>
+        <module>langchain4j-couchbase</module>
```
Looks like the modules are sorted alphabetically 😄

Heezer commented on docs/docs/integrations/embedding-stores/couchbase.md:
```diff
@@ -0,0 +1,154 @@
+
+# Couchbase
+
+https://www.trychroma.com/
```
wrong URL

chedim replied:
Can't believe I've missed that :D 
Thank you!

Heezer commented on langchain4j-couchbase/src/main/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStore.java:
```diff
@@ -0,0 +1,331 @@
+    }
+
+    @Override
+    public void removeAll(Collection<String> ids) {
```
inte test of removing (`extends EmbeddingStoreWithRemovalIT`) shall be added then. Remove by filter tests could be annotated with `@Ignore`

Heezer reviewed (changes requested):
Small fixes still needed

chedim committed:
CouchbaseEmbedding: Addresses PR comments; adds removal tests

chedim commented:
Ready for re-review.

chedim committed:
Update langchain4j-couchbase/src/main/java/dev/langchain4j/store/embedding/couchbase/CouchbaseEmbeddingStore.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

Heezer commented on langchain4j-couchbase/src/test/java/dev/langchain4j/store/embedding/couchbase/Utils.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.function.Supplier;
+
+@Testcontainers
+public final class Utils {
```
The ambiguous class name and everything public - is no go 😞  try an interface with default implementations instead

chedim replied:
made package-private and renamed the class to `CouchbaseTestUtils`, will that work?

Heezer replied:
not perfect, but fine for me 😄

Heezer reviewed (changes requested):
Utils must be refactored out

chedim committed:
Renames Utils into CouchbaseTestUtils and makes it package-private

Heezer reviewed (dismissed):
Thanks!!

Heezer committed:
Merge branch 'main' into main

chedim commented:
Looking into the tests errors, works fine on my machine ©

chedim committed:
Fixes integration tests

chedim committed:
Merge branch 'main' into main

chedim committed:
fixes overridden method in removal tests

chedim committed:
Merge branch 'main' into main

chedim commented:
Updated the branch, resolved conflict with main in docs.

chedim committed:
Merge branch 'main' into main

chedim committed:
Adapting to changes from main: Adds awaitility dependency for test scope, makes sure that tests work after the removal of waitUntilReady method

chedim committed:
Merge branch 'main' into main

chedim commented:
Updated the branch with main;
Added awaitility dependency for test scope;
Modified the tests and the storage to account for the removal of waitUntilReady method — the vector index was returning rows for documents that were yet to be persisted into the collection, causing an uncaught exception in the storage. Fixed now. 

Tests are passing locally.

chedim committed:
Merge branch 'main' into main

dliubarskyi committed:
Merge branch 'main' into main

chedim commented:
Tests are passing, the compliance workflow is passing for couchbase module and failing on Oracle.

chedim committed:
Merge branch 'main' into main

chedim committed:
Update docs/docs/integrations/embedding-stores/couchbase.md

Co-authored-by: LangChain4j <langchain4j@gmail.com>

chedim committed:
Update docs/docs/integrations/embedding-stores/couchbase.md

Co-authored-by: LangChain4j <langchain4j@gmail.com>

chedim committed:
Update docs/docs/integrations/embedding-stores/couchbase.md

Co-authored-by: LangChain4j <langchain4j@gmail.com>

chedim committed:
PR Comments

chedim commented on langchain4j-couchbase/src/main/java/dev/langchain4j/store/embedding/couchbase/Document.java:
```diff
@@ -0,0 +1,20 @@
+    private float[] vector;
+    private String text;
+    private Map<String, Object> metadata;
+    private double score;
```
It was needed for pulling documents directly from the index data, but that would be a future improvement so, for now I've removed it. Thank you for pointing out!

Heezer reviewed (dismissed):
Well done! 👍

dliubarskyi committed:
Update docs/docs/integrations/embedding-stores/couchbase.md
